### DISCUSSION TITLE
Different function calls for different Conditions

### DISCUSSION THREAD

**Discussion by trulybeardless** [skypjack/entt#885] [Q&A] [answered]
## Current Setup

I´ve got a simple Renderer which has 2 "Draw"-functions
Draw1 takes only a mesh as an argument and renders it with a default material
Draw2 takes a mesh **and** a material as an argument and renders it.
*(The function are both called Draw in the actual program, but for the sake of clarity I decided to call them Draw1 and Draw2)*

## Problem

Using the second function works just fine, I use a "view" to get both components and pass them onto the function.
The problem is I don´t know how check for just one component and then for the other.
### It would (in theory) work like this:
1. Check if entity has a mesh component
2. Check if entity has a material component
3. If an entity only has a mesh component, use Draw1
4. If an entity has both components, use Draw2

**How could I achieve that with ENTT?**

**Comment by codylico** [comment]
Maybe you could use a view that gives both entity identifier and mesh component, then try to look up the material by entity identifier? I'm not sure if and how that would work, though; maybe someone else can verify.

**Reply by Green-Sky** [reply]
basically. Use a view with all common components (mesh component) and inside the iteration check for conditional components (material).
Since you are provided with the entity(id) you can just `if (registry.all_of<material_component>(entity)) {} else {}`.

**Reply by Green-Sky** [reply]
There is an option to exclude specific components from a view, but I am not sure how much slower 2 separate view would be.

**Comment by skypjack** [comment, marked as answer]
If I get it correctly, this is what you're looking for?
```cpp
for(auto [entity, cmesh]: registry.view<mesh>(entt::exclude<material>).each()) { Draw1(cmesh); } // (1)
for(auto [entity, cmesh, cmaterial]: registry.view<mesh, material>().each()) { Draw2(cmesh. cmaterial); } // (2)
```
That is:
1. _If an entity only has a mesh component, use `Draw1`_ (and therefore it has not material component)
2. If an entity has both components, use `Draw2`

**Reply by trulybeardless** [reply]
Oh god, I can´t believe it is that simple, now I feel stupid!
Nevertheless: Thank you!

**Reply by skypjack** [reply]
You're welcome. 🙂

**Comment by codylico** [comment]
>
> The following is an idea I came up with:

Maybe use a view that gives both entity and mesh component, then try to
look up the material by entity ID? Someone else can confirm whether that
works....

**Reply by skypjack** [reply]
Yeah, it also works. It's mostly a matter of tastes I guess.
